// playwright-cli/lib/injected/consoleApi.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleAPI = void 0;
const selectorGenerator_1 = require("./selectorGenerator");
;
class ConsoleAPI {
    constructor(injectedScript) {
        this._injectedScript = injectedScript;
        window.playwright = {
            $: (selector) => this.querySelector(selector),
            $$: (selector) => this.querySelectorAll(selector),
            inspect: (selector) => this.inspect(selector),
            selector: (element) => this.buildSelector(element).selector,
        };
    }
    _checkSelector(parsed) {
        for (const { name } of parsed.parts) {
            if (!this._injectedScript.engines.has(name))
                throw new Error(`Unknown engine "${name}"`);
        }
    }
    querySelector(selector) {
        if (typeof selector !== 'string')
            throw new Error(`Usage: playwright.$('Playwright >> selector').`);
        const parsed = this._injectedScript.parseSelector(selector);
        this._checkSelector(parsed);
        const elements = this._injectedScript.querySelectorAll(parsed, document);
        return elements[0];
    }
    querySelectorAll(selector) {
        if (typeof selector !== 'string')
            throw new Error(`Usage: playwright.$$('Playwright >> selector').`);
        const parsed = this._injectedScript.parseSelector(selector);
        this._checkSelector(parsed);
        return this._injectedScript.querySelectorAll(parsed, document);
    }
    inspect(selector) {
        if (typeof window.inspect !== 'function')
            return;
        if (typeof selector !== 'string')
            throw new Error(`Usage: playwright.inspect('Playwright >> selector').`);
        window.inspect(this.querySelector(selector));
    }
    buildSelector(element) {
        return selectorGenerator_1.buildSelector(this._injectedScript, element);
    }
}
exports.ConsoleAPI = ConsoleAPI;
